Project/Assets/ML-Agents/Examples/Soccer/Scripts/AgentSoccer.cs

```diff
@@ -107,8 +140,22 @@ public void MoveAgent(float[] act)
 
     public override void OnActionReceived(float[] vectorAction)
     {
-        // Existential penalty for strikers.
-        AddReward(-1f / 3000f);
+
+        if (position == Position.Goalie)
+        {
+            // Existential bonus for Goalies.
+            AddReward(1f / 3000f);
+        }
+        else if (position == Position.Striker)
+        {
+            // Existential penalty for Strikers
+            AddReward(-1f / 3000f);
+        }
+        else
+        {
+            // Existential penalty cumulant for Generic
+            timePenalty += -1f / 3000f;
```

Comment:
In this way, I accumulate timestep penalty and give it all at once at the end of the episode _if an agent wins_. The point here is so that there's never additional utility in 'losing faster' i.e. if each agent got timestep penalty irrespective of outcome but still utility in winning faster.  The trained soccertwos agents look much more stable defensively and I attribute that to this.